subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptEvaluator.kt

```diff
@@ -207,4 +128,108 @@ class StandardKotlinScriptEvaluator(
+            kotlinScriptBasePluginsApplicator
+                .apply(project)
+        }
+
+        override fun closeTargetScopeOf(scriptHost: KotlinScriptHost<*>) {
+            val targetScope = scriptHost.targetScope
+//                    targetScope.export(classPathProvider.gradleApiExtensions)
+            pluginRequestsHandler.pluginRequestApplicator.applyPlugins(
+                DefaultPluginRequests.EMPTY,
+                scriptHost.scriptHandler as ScriptHandlerInternal?,
+                null,
+                targetScope)
+        }
+
+        override fun cachedClassFor(
+            templateId: String,
+            sourceHash: HashCode,
+            parentClassLoader: ClassLoader
+        ): Class<*>? =
+            classloadingCache
+                .get(cacheKeyFor(templateId, sourceHash, parentClassLoader))
+                ?.scriptClass
+
+        override fun cache(
+            templateId: String,
+            sourceHash: HashCode,
+            parentClassLoader: ClassLoader,
+            specializedProgram: Class<*>
+        ) {
+            classloadingCache.put(
+                cacheKeyFor(templateId, sourceHash, parentClassLoader),
+                LoadedScriptClass(
+                    CompiledScript(File(""), specializedProgram.name, Unit),
+                    specializedProgram))
+        }
+
+        private
+        fun cacheKeyFor(
+            templateId: String,
+            sourceHash: HashCode,
+            parentClassLoader: ClassLoader
+        ): ScriptCacheKey =
+            ScriptCacheKey(
+                templateId,
+                sourceHash,
+                parentClassLoader)
+
+        override fun cachedDirFor(
+            templateId: String,
+            sourceHash: HashCode,
+            parentClassLoader: ClassLoader,
+            initializer: (File) -> Unit
+        ): File =
+            try {
+                kotlinCompiler.cacheDirFor(cacheKeyPrefix + templateId + sourceHash.toString() + parentClassLoader) {
+                    initializer(baseDir)
+                }
+            } catch (e: CacheOpenException) {
+                throw e.cause as? ScriptCompilationException ?: e
+            }
+
+        private
+        val cacheKeyPrefix =
+            CacheKeyBuilder.CacheKeySpec.withPrefix("kotlin-dsl-interpreter")
+
+        override fun compilationClassPathOf(classLoaderScope: ClassLoaderScope): ClassPath =
+            classPathProvider.compilationClassPathOf(classLoaderScope)
+
+        override fun loadClassInChildScopeOf(
+            classLoaderScope: ClassLoaderScope,
+            childScopeId: String,
+            location: File,
+            className: String,
+            accessorsClassPath: ClassPath?
+        ): Class<*> =
+            classLoaderScope
+                .createChild(childScopeId)
+                .local(DefaultClassPath.of(location))
+                .apply { accessorsClassPath?.let(::local) }
+                .lock()
```

Comment:
Pre-mature locking of a `ClassLoaderScope` has caused a few issues. Just want to make sure those potential issues have been evaluated here.

#4823
#311

Reply:
Thanks. That's the same behaviour as before.